### PR TITLE
fix(es/minifier): Eliminate unused classes with cyclic references

### LINKED ISSUE #11934: [Minifier] Classes are not eliminated with cyclic references
### Describe the bug

See the playground

### Input code

```typescript
class A {
  method(){
    new B().method();
  }
}
class B extends A {
  method(){
    new A().method();
  }
}
```

### Config

```json
{
  "jsc": {
    "parser": {
      "syntax": "ecmascript",
      "jsx": false
    },
    "target": "es2024",
    "loose": false,
    "minify": {
      "compress": {
        "arguments": false,
        "arrows": true,
        "booleans": true,
        "booleans_as_integers": false,
        "collapse_vars": true,
        "comparisons": true,
        "computed_props": true,
        "conditionals": true,
        "dead_code": true,
        "directives": true,
        "drop_console": false,
        "drop_debugger": true,
        "evaluate": true,
        "expression": false,
        "hoist_funs": false,
        "hoist_props": true,
        "hoist_vars": false,
        "if_return": true,
        "join_vars": true,
        "keep_classnames": false,
        "keep_fargs": true,
        "keep_fnames": false,
        "keep_infinity": false,
        "loops": true,
        "negate_iife": true,
        "properties": true,
        "reduce_funcs": false,
        "reduce_vars": false,
        "side_effects": true,
        "switches": true,
        "typeofs": true,
        "unsafe": false,
        "unsafe_arrows": false,
        "unsafe_comps": false,
        "unsafe_Function": false,
        "unsafe_math": false,
        "unsafe_symbols": false,
        "unsafe_methods": false,
        "unsafe_proto": false,
        "unsafe_regexp": false,
        "unsafe_undefined": false,
        "unused": true,
        "const_to_let": true,
        "pristine_globals": true
      },
      "mangle": false
    }
  },
  "module": {
    "type": "es6"
  },
  "minify": true,
  "isModule": true
}
```

### Link to the code that reproduces this issue

https://play.swc.rs/?version=1.15.41&code=H4sIAAAAAAAAA0vOSSwuVnBUqOblUlDITS3JyE%2FR0ARzFBTyUssVnDQ09WDC1iDhWl4uIEoGa3NSSK0oSc1LwWeAIzYDALZpHTt2AAAA&config=H4sIAAAAAAAAA32US47bMAyG93OKwOsuikHRRQ%2FQXc8gKBLlKNXDEKlMjEHuXlq2k3RCe2fz409KJMXPt8OhO6Ppfh0%2B%2BZN%2FBl0Qyv2fLTgm0le2dGCiRlP8QN23lZ5xQk4HhGa6zaQjXXqgpsL37%2B8%2FFkUXckZYFYst%2BuTd%2BJzT5DgUQHyysZVD1giJ8H%2F9wkr%2BmACV%2Bmw%2F5hxApx2iNCqfCHooUmCTQ9ADgrroIkSZTqqLxyylmGAlsGooeRB5sp58TpzzlVrQVplsQUC%2BgCF%2FAUnGuViWkK8n3KdhC8fa963PX9Rw0aFqEnLCtbWETytEPWWPpFxNUglnuFGDGS7F%2Far0ThWgWtKr7px92ujJXwCuQNCISUeQ4jYPx%2FO0pXa7Sp8cjyyNAuf5lm6ZoOeiKu%2BdUNmpMlDIS90sYKuBqbJGOs6CN8qH3oIC53hWhND44cmcpKQ0DpCdALi%2F2klTNQN1f4UbfHoQO%2Fg335LkAVs8oqbTNsUxHnPYSRCBTtnuOHArKG%2FjwlviOmzzmizwaIAVXSo28LoE%2BAFQVqHty5fZ4OfBEVUf8vGxJhaH230PR536x3ufV%2FHb4tDFbGuDy5Kf%2Bjuv5p%2Fdw2ndwusJOo9%2FVmHLefsHO1REuy8GAAA%3D

### SWC Info output

_No response_

### Expected behavior

Both `A` and `B` should be eliminated, as terser does: 

Image

### Actual behavior

_No response_

### Version

1.15.41

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by baltasarblanco**
## Description

The minifier keeps classes that are only reachable through a cyclic reference to each other, even when the whole cluster is unused; `terser` removes them. Reproduced in #11934.

The bug is in the DCE tree-shaker (`swc_ecma_transforms_optimization::simplify::dce`), which the minifier runs through `perform_dce`.

### Root cause

`Data::subtract_cycles` cancels the internal edges of a strongly-connected component — making its members eliminable — only when no node in the component is a top-level entry and none has an incoming edge from outside it.

In `Analyzer::visit_class_decl` the `extends` clause was visited *before* the class identifier was pushed onto the dependency path (`with_ast_path`). A reference seen with an empty path is recorded as a top-level entry, so for `class B extends A {}` the superclass `A` was pinned as an entry. That kept the otherwise-unreachable `A ⇄ B` cycle (formed by the method bodies) from being subtracted, so both classes survived.

### Fix

`visit_class_decl` now attributes the `extends` edge to the subclass (inside the path) only when the superclass is a **backward reference to an already-initialized class**, tracked with a small per-scope `FxHashSet<Id>`. Forward references, non-identifier superclasses, and everything else keep the original behavior (visited outside the path, i.e. pinned as an entry). This preserves TDZ-observable cycles such as `class A extends B {} class B extends A {}` — which throws at definition time and must not be eliminated (there only `B → A` is a backward reference, so `B` stays pinned and both classes are kept).

It also closes a latent hole in the same mechanism: a clean class in a cycle that is referenced by a kept, side-effectful sibling (e.g. a `static {}` block) could be dropped, leaving a dangling reference. A class whose definition has side effects is now marked as an entry, reusing the existing class-drop predicate (factored out as `class_def_is_side_effect_free`), so those cycles are preserved.

### Performance

The change is confined to `dce/mod.rs` and is intentionally minimal: it does **not** touch the `VarInfo` edge-weight struct or the `subtract_cycles` hot loop. The only added work on the common path is one `FxHashSet<Id>` insert per class declaration. CodSpeed should be neutral. (An earlier, broader attempt — #11933 — grew the edge weight and added per-declarator bookkeeping, which regressed CodSpeed; this PR is scoped to the class-`extends` case to avoid that.)

### Tests

In `tests/simplify_dce.rs`:
- `class_extends_cycle_unused` — the repro; both classes are eliminated. Confirmed to **fail** without the fix (`git stash` on `dce/mod.rs`).
- `class_extends_cycle_tdz_preserved` — `class A extends B {} class B extends A {}` is kept.
- `class_extends_cycle_with_side_effect_preserved` — a cycle whose member has a `static {}` block is kept.

`cargo test -p swc_ecma_transforms_optimization` and `cargo test -p swc_ecma_minifier` pass with no snapshot changes (including the terser execution tests).

## Related issue

Closes #11934

### CODE REVIEW

**Comment by baltasarblanco** [comment]
Force-pushed to update one snapshot.

`Test - swc` was failing on `privateNamesConstructorChain-1.2.minified.js` (the `tsc` reference suite, minified variant). That fixture's input is two classes that only reference each other (`Parent.accessChildProps` does `new Child()`, `Child extends Parent`) with nothing instantiating or exporting either — i.e. the exact unreachable-cycle pattern this PR targets. With the fix the whole cluster is correctly eliminated under `minify`, leaving only the side-effect-only helper imports (which `preserve_imports_with_side_effects` keeps). I verified against the input that this is genuine dead-code elimination, not over-elimination, and regenerated the snapshot.

The local `swc_ecma_transforms_optimization`, `swc_ecma_minifier`, and `swc --test fixture/tsc` suites pass, and CodSpeed came back neutral.

**Comment by baltasarblanco** [comment]
Good catch — confirmed and fixed in a624e48.

You're right: `class_def_is_side_effect_free` treated the resolved identifier as a pure read, so a `static` initializer reading a not-yet-initialized class in the same SCC let `subtract_cycles` drop both, silently removing the required TDZ throw.

The fix marks a class as an entry when a **`static`** field initializer or `static {}` block reads an identifier not yet recorded in `initialized_classes` at definition time — i.e. a definition-time TDZ read. The check runs before the current class is inserted into that set, so in `class A { static x = B } class B extends A {}`, `A` correctly sees `B` as uninitialized and is preserved.

This stays narrow: only `static` members evaluate at definition time. Instance fields and method bodies (including this PR's target case, where the cross-references live in method bodies) run at instantiation/call time, are not TDZ-observable, and remain eliminable.

Added `class_static_field_reads_later_class_tdz` as a regression test — it fails without the guard (both classes dropped) and passes with it. Full `swc_ecma_transforms_optimization` suite and clippy are clean.

**Comment by baltasarblanco** [comment]
Thanks for the thorough review — I worked through all four cases. Summary in 3d7afdd:

**`top_retain` and decorated-superclass cycles — fixed (these were real regressions this PR would introduce).**
I checked against `main` using a clean worktree: `main` *preserves* both
`@dec class A { … B … } class B extends A { … A … }` and the `top_retain: ["A"]` cycle.
The backward-edge optimization would have collapsed them. The guard now requires,
beyond the superclass being an already-initialized identifier: the subclass has no
decorators **and** the superclass is not decorated (tracked in a `decorated_classes`
set), and neither class is in `top_retain`. Regression tests
`class_extends_cycle_decorated_super_preserved` and
`class_extends_cycle_top_retain_preserved` both fail without the guard.

**Static-field TDZ — fixed (earlier commit).**
`class A { static x = B } class B extends A {}` is preserved via
`static_init_reads_uninitialized`; covered by `class_static_field_reads_later_class_tdz`.

**Switch-case TDZ — pre-existing on `main`, not introduced by this PR.**
`switch (x) { case 1: class A {…B…} case 2: class B extends A {} }` is already
eliminated by `main` today, independently of this change — I confirmed it by running
the same input on a clean `main` checkout, where both classes are dropped without any
of this PR's code. So this PR neither introduces nor worsens that case. Properly
fixing the switch/TDZ interaction would require modeling non-linear control flow in
the tree-shaker, which seems out of scope for an `extends`-cycle fix; happy to open a
separate issue for it if that's useful.

The full `swc_ecma_transforms_optimization` and `swc` fixture/tsc suites pass, and
CodSpeed is neutral.

**Comment by baltasarblanco** [comment]
Good catch — fixed in 41a9aaf.

The static-init TDZ scan (`static_init_reads_uninitialized`) was indeed missing
`ClassMember::PrivateProp`. A `static` private field is evaluated at definition
time just like a public one, so `class A { static #x = B } class B extends A {}`
read `B` in the TDZ but the scan didn't see it, letting the cycle be subtracted.

The scan now covers all four static member kinds that can carry an initializer:
`ClassProp`, `PrivateProp`, `AutoAccessor`, and `StaticBlock`. Added
`class_static_private_field_reads_later_class_tdz` as a regression test — it fails
without the added arm. Full `swc_ecma_transforms_optimization` suite and clippy clean.

**Comment by baltasarblanco** [comment]
After several rounds of definition-time edge cases from the review (TDZ via static initializers including private fields, decorators, nested class heritage), I changed the guard's strategy in a64c91b.

Instead of enumerating each hazard, the backward-reference optimization now applies only to **structurally-trivial classes on both sides of the `extends` edge** — classes whose definition evaluates nothing eagerly except the heritage reference itself: no decorators (on the class or any member), and no `static` initializers or `static {}` blocks. Anything else falls back to the original behavior and is preserved. This is a single structural predicate (`class_def_is_trivial`) that replaced the earlier per-case scans and auxiliary set, so the code is a bit smaller too.

The tradeoff is that this is **deliberately narrow** — more conservative than terser, which eliminates these cycles even when they would throw at definition time. We only collapse cycles of plain-method classes, which is exactly the reported case (#11934). I checked the eliminated cases against terser and confirmed the preserved cases are genuine definition-time evaluations.

I'm very open to taking this in whatever direction you prefer — a more aggressive version closer to terser, a different framing of the guard, widening or narrowing the scope, or if you feel the optimization isn't worth the added complexity at all. Happy to rework or drop it however makes most sense for the codebase. Just let me know what you'd like.

**Comment by baltasarblanco** [comment]
@kdy1 — before I put more work into this, I'd appreciate your read on whether it's worth pursuing.

The automated review has surfaced a steady stream of same-family edge cases: definition-time TDZ through static initializers (including private fields), class/member/parameter decorators, and cycles that stay observable through a decorated function or a `top_retain`'d sibling. The current commit guards the two endpoints of the `extends` edge, but several of these hazards live on *other* members of the cycle, where an endpoint check can't reach them.

There's a cleaner structural direction available: instead of guarding the edge, pin every must-preserve node (decorated, `top_retain`'d, or with definition-time evaluation) as a graph entry and let the existing whole-cycle check in `subtract_cycles` preserve any SCC that touches one. That folds the per-case guards and the auxiliary set into a single mechanism and closes the family by construction — but I haven't implemented it yet.

Given that the win here is narrow (eliminating mutually-referencing unused classes, the reported case), it's completely reasonable if you'd rather this stay conservative, go a different way, or not land at all. I'm happy to implement the structural version, keep the current one, or close the PR — whichever you think is the right complexity/benefit trade-off for the tree-shaker. No rush.

**Comment by kdy1** [comment]
I think the problem is that we don't have enough documentation about TDZ in `AGENTS.md` files. The minifier does not preserve TDZ, so those TDZ reviews are meaningless.

I made https://github.com/swc-project/swc/pull/11966
Once it's merged, Codex will be able to do some better review

File: crates/swc_ecma_transforms_optimization/src/simplify/dce/mod.rs [hunk 1]
```diff
@@ -468,12 +563,53 @@ impl Visit for Analyzer<'_> {
...
+            Expr::Ident(i) => Some(i.to_id()),
+            _ => None,
+        });
+        let super_is_initialized_class_ref = class_def_is_trivial(&n.class)
+            && !self.config.top_retain.contains(&n.ident.sym)
+            && super_id.as_ref().is_some_and(|id| {
+                self.initialized_classes.contains(id)
+                    && !self.nontrivial_classes.contains(id)
+                    && !self.config.top_retain.contains(&id.0)
+            });
```
[the remark this answers is not in the document]
Comment (kdy1) [reply crates/swc_ecma_transforms_optimization/src/simplify/dce/mod.rs:598]: This sounds legit. We need to check for direct evals

[the remark this answers is not in the document]
Comment (kdy1) [reply crates/swc_ecma_transforms_optimization/src/simplify/dce/mod.rs:598]: Well. I'm not sure if DCE is even invoked for minifier when there's a direct eval

[the remark this answers is not in the document]
Comment (baltasarblanco) [reply crates/swc_ecma_transforms_optimization/src/simplify/dce/mod.rs:598]: Checked the code: DCE does run with a direct eval. `perform_dce` is only gated on `unused`/`dead_code`, not on eval, so the tree-shaker runs regardless.

The tree-shaker does handle eval, but only inside function scopes (`bindings_affected_by_eval` via `collect_decls`). There's no equivalent at module top level, so a top-level `eval()` sets `found_direct_eval` without marking the top-level bindings as used. That's reachable exactly when `top_level` is enabled — which is also the only case where this optimization drops top-level classes — so the report is real, and it's a regression introduced here: previously the heritage reference pinned the superclass as a top-level entry and kept the cycle alive.

I can fix it by mirroring the function-scope handling at the top level: when a direct eval is found in the root scope, treat top-level declarations as used (pin them as entries) so any cycle they belong to is preserved. Want me to go that way, or would you prefer something narrower — e.g. just skip the internal-edge path when a top-level eval is present?

**Review by kdy1 — DISMISSED** [review]
Thank you so much for the contribution! Actaully this issue was a headache

**Review by kdy1 — APPROVED** [review]
